storage/remote/graphite/client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graphite
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"net"
+	"net/url"
+	"sort"
+	"time"
+
+	"github.com/prometheus/common/log"
+
+	"github.com/prometheus/common/model"
+)
+
+// Client allows sending batches of Prometheus samples to Graphite.
+type Client struct {
+	url     string
+	timeout time.Duration
+	prefix  string
+}
+
+// NewClient creates a new Client.
+func NewClient(url string, timeout time.Duration, prefix string) *Client {
+	return &Client{
+		url:     url,
+		timeout: timeout,
+		prefix:  prefix,
+	}
+}
+
+func pathFromMetric(m model.Metric, prefix string) string {
+	buffer := bytes.NewBufferString("")
+
+	buffer.WriteString(prefix)
+	buffer.WriteString(escape(m[model.MetricNameLabel]))
+
+	// We want to sort the labels.
+	labels := make(model.LabelNames, len(m))
+	for l, _ := range m {
+		labels = append(labels, l)
+	}
+	sort.Sort(labels)
+
+	// For each label, in oder, add ".<label>.<value>".
+	for _, l := range labels {
+		v := m[l]
+
+		if l == model.MetricNameLabel || len(l) == 0 {
+			continue
+		}
+		// Here we use '=' instead of '.' to be able
+		// to later read back the value correctly. Using '.' would
+		// not allow to distinguish labels from values.
+		buffer.WriteString(fmt.Sprintf(
+			".%s=%s", string(l), escape(v)))
+	}
+	return string(buffer.Bytes())
+}
+
+// Store sends a batch of samples to Graphite.
+func (c *Client) Store(samples model.Samples) error {
+	u, err := url.Parse(c.url)
+	if err != nil {
+		return err
+	}
+
+	conn, err := net.DialTimeout(u.Scheme, u.Host, c.timeout)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	buf := bytes.NewBufferString("")
+	for _, s := range samples {
+		k := pathFromMetric(s.Metric, c.prefix)
+		t := s.Timestamp.Unix()
+		v := float64(s.Value)
+		if math.IsNaN(v) || math.IsInf(v, 0) {
+			log.Warnf("cannot send value %f to Graphite,"+
```

Comment:
I think both in the OpenTSDB writer and in the Graphite one, we should stop logging this line because it creates a lot of noise (there are situations where it's super common to get special float values). We should increment a metric about this instead, like we do in the InfluxDB writer: https://github.com/prometheus/prometheus/blob/master/storage/remote/influxdb/client.go#L73

But IMO it's fine to change that in a later PR.